### PR TITLE
Add experimental exportTrailingSlash config

### PR DESCRIPTION
**Pull request by ijjk**
By default when `next export`ing a Next.js application we will automatically append a `/` to all urls to be fully compatible with the directory structure being output.

However since most platforms support directory indexes it makes sense to change this default in the future.

This PR adds `exportTrailingSlash` as experimental flag. We'll try this out for a bit on nextjs.org / zeit.co/docs before introducing it as new option.

The default value is `true` as this is the current behavior in stable Next.js.

```
{
  experimental: {
    exportTrailingSlash: false
  }
}
```

⚠️ as with all experimental flags being added this is subject to breaking between canary/stable versions.

### CODE REVIEW

File: packages/next/client/link.js [hunk 1]
```diff
@@ -42,16 +42,23 @@ class Link extends Component {
     }
   }
 
+  formatUrl (url, asHref) {
```
Comment (timneutkens) [inline packages/next/client/link.js]: Since this method only uses static methods I'd move it out of the class

File: packages/next/client/link.js [hunk 2]
```diff
@@ -42,16 +42,23 @@ class Link extends Component {
...
+    url = url && typeof url === 'object'
+      ? formatWithValidation(url)
+      : url
+
+    if (url && asHref && process.env.__NEXT_EXPORT_TRAILING_SLASH) {
```
Comment (timneutkens) [inline packages/next/client/link.js]: It's generally best to wrap it like this:

```
if(process.env.__NEXT_EXPORT_TRAILING_SLASH) {
  if(url && asHref) {//etc}
}
```

File: packages/next-server/lib/router/router.ts [hunk 3]
```diff
@@ -206,7 +206,7 @@ export default class Router implements IRouterInterface {
       // Add the ending slash to the paths. So, we can serve the
       // "<page>/index.html" directly for the SSR page.
       // @ts-ignore this is temporarily global (attached to window)
-      if (__NEXT_DATA__.nextExport) {
+      if (!process.env.__NEXT_EXPORT_TRAILING_SLASH && __NEXT_DATA__.nextExport) {
```
Comment (timneutkens) [inline packages/next-server/lib/router/router.ts]: Same as https://github.com/zeit/next.js/pull/6664/files#r265789907

Shouldn't this condition be without a `!`. If it's true it has to add the trailing slash right.

File: packages/next/client/link.js (hunk 1 above)
Comment (ijjk) [reply packages/next/client/link.js]: Good point, updated

File: packages/next/client/link.js (hunk 2 above)
Comment (ijjk) [reply packages/next/client/link.js]: Updated

File: packages/next-server/lib/router/router.ts (hunk 3 above)
Comment (ijjk) [reply packages/next-server/lib/router/router.ts]: Right, updated

File: packages/next/client/link.js [hunk 4]
```diff
@@ -23,14 +23,27 @@ function memoizedFormatUrl (formatUrl) {
...
     return result
   }
 }
 
+function formatUrl (url, isAs) {
+  url = url && typeof url === 'object'
+    ? formatWithValidation(url)
+    : url
+
+  if (process.env.__NEXT_EXPORT_TRAILING_SLASH) {
```
Comment (timneutkens) [inline packages/next/client/link.js]: Why is this located here 🤔 It's already there on line 140 and just needs to be wrapped in the if statement.

Comment (timneutkens) [reply packages/next/client/link.js]: This causes the wrong result btw, it's appending `/` on non-export pages.

File: test/integration/basic/test/client-navigation.js [hunk 5]
```diff
@@ -467,7 +467,7 @@ export default (context) => {
         expect(text).toBe('10')
 
         expect(await browser.url())
-          .toBe(`http://localhost:${context.appPort}/nav/querystring/10#10`)
+          .toBe(`http://localhost:${context.appPort}/nav/querystring/10/#10`)
```
Comment (timneutkens) [inline test/integration/basic/test/client-navigation.js]: These tests shouldn't be changed, the code in this PR is adding the wrong behavior hence why these failed.

File: packages/next/client/link.js (hunk 4 above)
Comment (ijjk) [reply packages/next/client/link.js]: Oh gotcha 🤦‍♂️

File: test/integration/basic/test/client-navigation.js (hunk 5 above)
Comment (ijjk) [reply test/integration/basic/test/client-navigation.js]: Reverted

**Review by timneutkens — CHANGES_REQUESTED** [review]
I think we should add tests for this new option 🤔

**Comment by ijjk** [comment]
Added a test with the option disabled

**Review by timneutkens — APPROVED** [review]
Amazing work @ijjk 💯🚀

**Review by timneutkens — APPROVED** [review]
Amazing work @ijjk 💯🚀

**Pull request merged**